### ISSUE TITLE
k8s Provider: use networking.k8s.io/v1 API

### ISSUE THREAD

**Issue by phin1x** [traefik/traefik#7466]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

With Kubernetes 1.19 the Ingress API and Ingress Class API  ist now stable (graduated to networking.k8s.io/v1) and will be removed from the networking.k8s.io/v1beta1 and extensions/v1beta1 API Group with Version 1.22.

Traefik still receive ingress objects via the extensions/v1beta1 informer:
https://github.com/traefik/traefik/blob/015f24a90166163160819c0fb3dcfc50c44aeaba/pkg/provider/kubernetes/ingress/client.go#L150
Traefik should check the cluster version and use the stable networking.k8s.io/v1 ingress and ingress class api if supported (1.19+) or the older extensions/v1beta1 and networking.k8s.io/v1beta1 API if the cluster version is blow 1.19 and convert them to networking.k8s.io/v1 objects.
Since this is already implemented (version check and conversion)  it should not take much effort.

**Comment by SantoDE** [comment]
Hello @phin1x,

thanks for your interest.

Why do you think that change is required? With our current mechanism we are fully backward compatible and can fetch all ingresses, no matter the version.

What is the issue with that approach? Do you see a problematic edge case?

**Comment by phin1x** [comment]
Hi @SantoDE ,

From the kubernetes 1.16 release notes:

Ingress resources will no longer be served from extensions/v1beta1 in v1.20. Migrate use to the networking.k8s.io/v1beta1 API, available since v1.14. Existing persisted data can be retrieved via the networking.k8s.io/v1beta1 API.

So traefik will not work with the next version of kubernetes. This will prevent us from upgrading kubernetes to the next minor release in january 2021.
The extension/v1beta1 api has been deprecated for over 1.5 years and should no longer be used.

**Comment by phin1x** [comment]
The release notes seams to be inconsistent, the removal of the extensions/v1beta1 is scheduled for kubernetes 1.22 which will be released in about one year. With this release, only networking.k8s.io/v1 will be servered, so a change is required anyway.

The timeline can be found here: https://github.com/kubernetes/kubernetes/issues/43214

**Comment by ldez** [comment]
https://github.com/traefik/traefik/blob/015f24a90166163160819c0fb3dcfc50c44aeaba/pkg/provider/kubernetes/ingress/client.go#L211

https://github.com/traefik/traefik/blob/015f24a90166163160819c0fb3dcfc50c44aeaba/pkg/provider/kubernetes/ingress/client.go#L177

**Comment by traefiker** [comment]
Hi! I'm Træfiker :robot: the bot in charge of tidying up the issues.

I have to close this one because of its lack of activity :disappointed:

Feel free to re-open it or join our [Community Forum](https://community.containo.us/).

**Comment by traefiker** [comment]
Hi! I'm Træfiker :robot: the bot in charge of tidying up the issues.

I have to close this one because of its lack of activity :disappointed:

Feel free to re-open it or join our [Community Forum](https://community.containo.us/).

**Comment by traefiker** [comment]
Hi! I'm Træfiker :robot: the bot in charge of tidying up the issues.

I have to close this one because of its lack of activity :disappointed:

Feel free to re-open it or join our [Community Forum](https://community.containo.us/).

**Comment by Gallardo994** [comment]
Could be related: cannot use ingressClass in helm installation due to missing networking.k8s.io/v1beta1 (Kubernetes 1.19.3-do2 in DigitalOcean DOKS)

ERROR: You must have atleast networking.k8s.io/v1beta1 to use ingressClass

**Comment by mikesir87** [comment]
Curious of the status of this, as it seems to have stalled a bit. Some of our Helm charts are starting to deploy Ingress objects from `networking.k8s.io/v1` to avoid deprecation warnings (assuming the cluster supports it), but then ingress handling is breaking because Traefik doesn't support it yet.

**Comment by notsureifkevin** [comment]
hey @mikesir87, based on https://github.com/traefik/traefik/pull/7549 it is scheduled to be released in the 2.4 GA (which is sometime next month). We are currently preparing an RC release, but I'm unsure as to whether or not it'll make it in time for the first revision.

**Comment by mikesir87** [comment]
Ok! Thanks @notsureifkevin! I saw the PR, but saw there hadn't been any activity on it for a few weeks. Looking forward to the next release then 😄

**Comment by phin1x** [comment]
Hi, will the pr make it into the upcoming 2.4 release? I saw a first release candidate, but the pr was not included.

**Comment by SantoDE** [comment]
It will probably be in 2.5 @phin1x.

It should not be a real issue tough, as Ingress from `networking/v1` are working already. Its basically an internal architecture change only.

**Comment by danieldaeschle** [comment]
If i use `networking.k8s.io/v1` i get this:
```
error: error validating "hello-kubernetes.yaml": error validating data: [ValidationError(Ingress.spec.rules[0].http.paths[0].backend): unknown field "serviceName" in io.k8s.api.networking.v1.IngressBackend, ValidationError(Ingress.spec.rules[0].http.paths[0].backend): unknown field "servicePort" in io.k8s.api.networking.v1.IngressBackend]; if you choose to ignore these errors, turn validation off with --validate=false
```

**Comment by phin1x** [comment]
@danieldaeschle the Ingress API changed from v1beta1 to v1:

v1beta1:
```
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: minimal-ingress-v1beta1
spec:
  rules:
  - http:
      paths:
      - path: /testpath
        pathType: Prefix
        backend:
          serviceName: test
          servicePort: 80
```

v1:
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: minimal-ingress-v1
spec:
  rules:
  - http:
      paths:
      - path: /testpath
        pathType: Prefix
        backend:
          service:
            name: test
            port:
              number: 80
```

Here you find the current documentation: https://kubernetes.io/docs/concepts/services-networking/ingress/

**Comment by danieldaeschle** [comment]
Thanks!

**Comment by danieldaeschle** [comment]
As you can see in my comment i already use `networking.k8s.io/v1`. I still get this error.

**Comment by phin1x** [comment]
@danieldaeschle it seems that you are using the spec of the old v1beta version but set the new v1 version as the target.
you have to convert your ingress definition between the two versions. in the new v1 version, the serviceName and servicePort keys in the backend definion does not exists anymore (look at my example).

**Comment by phin1x** [comment]
closed, pr #7549 is merged
